src/Solvers/batched_tridiagonal_solver.jl

```diff
@@ -73,48 +70,132 @@ Reference implementation per Numerical Recipes, Press et. al 1992 (§ 2.4).
+        f₁ = get_coefficient(f, m, 1, n, grid, p, tridiagonal_direction, args...)
+        ϕ[m, 1, n] = f₁ / β
+
+        @unroll for q = 2:N
+            cᵏ⁻¹ = get_coefficient(c, m, q-1, n, grid, p, tridiagonal_direction, args...)
+            bᵏ   = get_coefficient(b, m, q,   n, grid, p, tridiagonal_direction, args...)
+            aᵏ⁻¹ = get_coefficient(a, m, q-1, n, grid, p, tridiagonal_direction, args...)
 
-        @unroll for k = 2:Nz
-            cᵏ⁻¹ = get_coefficient(c, i, j, k-1, grid, p, args...)
-            bᵏ   = get_coefficient(b, i, j, k,   grid, p, args...)
-            aᵏ⁻¹ = get_coefficient(a, i, j, k-1, grid, p, args...)
+            t[m, q, n] = cᵏ⁻¹ / β
+            β = bᵏ - aᵏ⁻¹ * t[m, q, n]
 
-            t[i, j, k] = cᵏ⁻¹ / β
-            β = bᵏ - aᵏ⁻¹ * t[i, j, k]
+            fᵏ = get_coefficient(f, m, q, n, grid, p, tridiagonal_direction, args...)
 
-            fᵏ = get_coefficient(f, i, j, k, grid, p, args...)
-            
             # If the problem is not diagonally-dominant such that `β ≈ 0`,
             # the algorithm is unstable and we elide the forward pass update of ϕ.
             definitely_diagonally_dominant = abs(β) > 10 * eps(float_eltype(ϕ))
             !definitely_diagonally_dominant && break
-            ϕ[i, j, k] = (fᵏ - aᵏ⁻¹ * ϕ[i, j, k-1]) / β
+            ϕ[m, q, n] = (fᵏ - aᵏ⁻¹ * ϕ[m, q-1, n]) / β
         end
 
-        @unroll for k = Nz-1:-1:1
-            ϕ[i, j, k] -= t[i, j, k+1] * ϕ[i, j, k+1]
+        @unroll for q = N-1:-1:1
+            ϕ[m, q, n] -= t[m, q+1, n] * ϕ[m, q+1, n]
         end
     end
 end
 
+@kernel function solve_batched_tridiagonal_system_kernel!(ϕ, a, b, c, f, t, grid, p, args, tridiagonal_direction::Val{:z})
+    N = size(grid, 3)
+    m, n = @index(Global, NTuple)
+
+    @inbounds begin
+        β  = get_coefficient(b, m, n, 1, grid, p, tridiagonal_direction, args...)
+        f₁ = get_coefficient(f, m, n, 1, grid, p, tridiagonal_direction, args...)
+        ϕ[m, n, 1] = f₁ / β
+
+        @unroll for q = 2:N
+            cᵏ⁻¹ = get_coefficient(c, m, n, q-1, grid, p, tridiagonal_direction, args...)
+            bᵏ   = get_coefficient(b, m, n, q,   grid, p, tridiagonal_direction, args...)
+            aᵏ⁻¹ = get_coefficient(a, m, n, q-1, grid, p, tridiagonal_direction, args...)
+
+            t[m, n, q] = cᵏ⁻¹ / β
+            β = bᵏ - aᵏ⁻¹ * t[m, n, q]
+
+            fᵏ = get_coefficient(f, m, n, q, grid, p, tridiagonal_direction, args...)
+
+            # If the problem is not diagonally-dominant such that `β ≈ 0`,
+            # the algorithm is unstable and we elide the forward pass update of ϕ.
+            definitely_diagonally_dominant = abs(β) > 10 * eps(float_eltype(ϕ))
+            !definitely_diagonally_dominant && break
+            ϕ[m, n, q] = (fᵏ - aᵏ⁻¹ * ϕ[m, n, q-1]) / β
+        end
+
+        @unroll for q = N-1:-1:1
+            ϕ[m, n, q] -= t[m, n, q+1] * ϕ[m, n, q+1]
+        end
+    end
+end
```

Comment:
This looks great! However, I would like these to be rewritten using `i, j, k`. While it is easier to copy/paste with `m, n, q` I think it will be easier to read with `i, j, k` as we are used to. Can you please rewrite that way?

Reply:
Done